GitUI/CommandsDialogs/FormBrowse.cs

```diff
@@ -2344,51 +2300,77 @@ private void CommandsToolStripMenuItem_DropDownOpening(object sender, EventArgs
+            // "Pull/Fetch..." menu item always opens the dialog
+            DoPull(false, pullAction: AppSettings.DefaultPullAction);
+        }
 
-            public ScopedSetLastPullAction(AppSettings.PullAction newPullAction)
-            {
-                _oldPullAction = AppSettings.DefaultPullAction;
-                AppSettings.DefaultPullAction = newPullAction;
-            }
+        private void ToolStripButtonPullClick(object sender, EventArgs e)
+        {
+            // Clicking on the Pull button toolbar button will perform the default selected action silently,
+            // except if that action is to open the dialog (PullAction.None)
+            bool isSilent = AppSettings.DefaultPullAction != AppSettings.PullAction.None;
+            DoPull(isSilent: isSilent, pullAction: AppSettings.DefaultPullAction);
+        }
 
-            void IDisposable.Dispose()
-            {
-                AppSettings.DefaultPullAction = _oldPullAction;
-            }
+        private void pullToolStripMenuItem1_Click(object sender, EventArgs e)
+        {
+            // "Open Pull Dialog..." toolbar menu item always open the dialog with the current default action
+            DoPull(isSilent: false, pullAction: AppSettings.DefaultPullAction);
         }
 
         private void mergeToolStripMenuItem_Click(object sender, EventArgs e)
         {
-            using (new ScopedSetLastPullAction(AppSettings.PullAction.Merge))
-            {
-                PullToolStripMenuItemClick(sender, e);
-            }
+            DoPull(isSilent: true, pullAction: AppSettings.PullAction.Merge);
         }
 
         private void rebaseToolStripMenuItem1_Click(object sender, EventArgs e)
         {
-            using (new ScopedSetLastPullAction(AppSettings.PullAction.Rebase))
-            {
-                PullToolStripMenuItemClick(sender, e);
-            }
+            DoPull(isSilent: true, pullAction: AppSettings.PullAction.Rebase);
         }
 
         private void fetchToolStripMenuItem_Click(object sender, EventArgs e)
         {
-            using (new ScopedSetLastPullAction(AppSettings.PullAction.Fetch))
-            {
-                PullToolStripMenuItemClick(sender, e);
-            }
+            DoPull(isSilent: true, pullAction: AppSettings.PullAction.Fetch);
         }
 
-        private void pullToolStripMenuItem1_Click(object sender, EventArgs e)
+        private void fetchAllToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            DoPull(isSilent: true, pullAction: AppSettings.PullAction.FetchAll);
+        }
+
+        private void fetchPruneAllToolStripMenuItem_Click(object sender, EventArgs e)
         {
-            using (new ScopedSetLastPullAction(AppSettings.PullAction.None))
+            DoPull(isSilent: true, pullAction: AppSettings.PullAction.FetchPruneAll);
+        }
+
+        private void DoPull(bool isSilent, AppSettings.PullAction pullAction)
```

Comment:
Suggest reordering the args to be more logical: `(AppSettings.PullAction pullAction, bool isSilent)`.